### ISSUE TITLE
Precedence of `|>`

### ISSUE THREAD

**Issue by nicolo-ribaudo** [js-choi/proposal-hack-pipes#11]
On the left, `|>` is as tight as assignment operators. This allows parsing `x = y |> z` as `x = (y |> z)`, and `x ? y : z |> w` as `(x ? y : z) |> w`.

On the right, there are two options:
1. `|> PipeExpression`
2. `|> AssignmentExpression`

(2) matches the precedence of `=>`, which is the most similar operator to `|>` (they both take something on their left and put it in the expression on their right). It also allows thinks like `x |> res = %` and `x |> () => %`. However when used in longer chains, `x |> y = % |> f(%)` would be parsed as `x |> (y = % |> f(%))` and  not as `x |> (y = %) |> f(%)`.

(1) disallows unparenthesized assignments and arrows on the right of `|>`.

**Comment by theScottyJam** [comment]
I think, more often than not, the pipeline operator is going to be used over multiple lines. With that in mind:

```javascript
let temp
const x = data
  |> temp = doStuff(%)
  |> g()
```

I don't think you would expect that assignment to also grab the `|> g()` from the following lines - for this reason, I would rather force parentheses (option 1) rather than go with option 2.

**Comment by ljharb** [comment]
ick, why is doing an assignment inside a pipe even allowed in the first place? (i know the answer, because JS allows assignment in expression positions, sadly)

that's indeed what should require forced parens, since nobody should be doing that anyways and linters and styleguides will ban it regardless

**Comment by theScottyJam** [comment]
Would it be possible to make it so if you're currently on the RHS of a pipeline operator, and you've entered the body of an arrow function, and another "|>" token is encountered, then you auto-close the arrow function body? This sort of makes "|>" act like delimiters, separating out one expression from the next. This would mean the following:

```javascript
const result = data
  |> x => % + x
  |> y => %() + y
// same as
const result = data
  |> (x => % + x) // Note how this function definition got auto-closed
  |> y => %() + y

const fn = x => a |> % + 1
// same as
const fn = x => (a |> % + 1) // This "|>" token was the first one found, so it didn't auto close the arrow function.
```

This sort of thing would be more useful for F#, but is still applicable here. I don't see any issues with this yet ...

**Comment by ljharb** [comment]
```js
a |> b(c |> d) |> e
```

**Comment by theScottyJam** [comment]
Good call. I think the concept should still be possible, I'm just not sure the best way to formulate it in a rigorous way.

But, the gist of the idea is to try and treat "|>" similar to how commas are treated when separating function parameters, i.e. anything between two "|>" in the same expression should be treated as one unit. i.e. `a |> whatever... |> c` is the same as `a |> (whatever...) |> c`, just like `f(a, whatever..., c)` is the same as `f(a, (whatever...), c)`. Just like you can nest function calls, and the comma delimiter still works properly (e.g. `f(a, b(c, d), e)`), likewise, you should be able to nest pipelines, when things like parentheses get involved.

Does that make any sense?

**Comment by js-choi** [comment]
> ick, why is doing an assignment inside a pipe even allowed in the first place? (i know the answer, because JS allows assignment in expression positions, sadly)
> 
> that's indeed what should require forced parens, since nobody should be doing that anyways and linters and styleguides will ban it regardless

We could do the same thing we’re doing with `yield` expressions and make unparenthesized assignment in pipe bodies an early error, forcing the user to parenthesize. 

I could edit the explainer and spec with this change. How does that sound?

**Comment by nicolo-ribaudo** [comment]
Note that `AssignmentExpression` is:
- `yield` (disallowed)
- actual assignments (maybe disallowed)
- arrow functions
- conditional expressions

So the decision becomes: do we want to allow things like
```js
const makeAdderLogger = add => add
                                |> num => % + num |> console.log(%)

const add2AndLog = makeAdderLogger(2);

add2AndLog(4); // logs "6"
```

or not.

**Comment by theScottyJam** [comment]
For clarification @nicolo-ribaudo - would your code example be interpreted as this?
```javascript
const makeAdderLogger = add => (add
                                |> num => ((% + num) |> console.log(%))
)
```

And the decision is, whether or not your example should be allowed without being forced to put parentheses around the function definitions?

How would the following be interpretted?
```javascript
const result = data
  |> x => x + %
  |> console.log(%)
```

Like this?

```javascript
const result = data
  |> x => (x + %
  |> console.log(%))
```

If so, we should definitely force parentheses around function definitions.

**Comment by nicolo-ribaudo** [comment]
Yes, both your interpretations are correct.

**Comment by js-choi** [comment]
We can view this situation as an analogue of confusing grouping when we combine `||` and `??`. 

`??` functionally has the same precedence level as `||` (and `&&`). However, `??` plus `||` or `&&` cannot be used together without parentheses, in order to prevent confusion like with `a ?? b || c ?? d`. Early errors enforcing this exist in the clauses for `??`, `||`, and `&&`.

Likewise, we could prohibit (from both the left- and right-hand sides of `|>`) all unparenthesized expressions that are at the same precedence level of `|>`:

“`|>` functionally has the same precedence level as `=`, `=>`, `?` `:`, and `yield`. However, `|>` plus `=>`, `?` `:`, or `yield` cannot be used together without parentheses, in order to prevent confusion like with `a |> b => c(%) |> d(%)`. Early errors enforcing this exist in the clauses for `|>`, `=`, `=>`, `?` `:`, and `yield`.” This change would hopefully address #12 too.

However, I still want to try to preserve having `a => b |> c(%)` be equivalent to `a => (b |> c(%))` – and having `a |> b ? c(%) : d(%)` be equivalent to `a |> (b ? c(%) : d(%))`…I am not yet sure how best to do this.

**Comment by js-choi** [comment]
After more thought, I think we should require all conditionals to be parenthesized in pipe heads/bodies after all.

I’m thinking that the most desirable behavior might be to ban any any other assignment-level operation without parentheses as a pipe head or body. (The other assignment-level operations are `=>`, `=`, `+=`, etc., `?` `:`, and `yield`.)

I plan to edit the spec and explainer with this change.

<table>
<tr>
<tr><td colspan="2"><strong>Arrows</strong>

<tr>
<td><code>a => b |> %.c |> %.d</code>

<td><code>a => (b |> %.c |> %.d)</code>

<tr>
<td><code>a |> (b => %.c) |> %.d</code>

<td><code>a |> (b => %.c) |> %.d</code>

<tr>
<td><code>a |> (b => %.c |> %.d)</code>

<td><code>a |> (b => (%.c |> %.d))</code>

<tr>
<td><code>a |> b => %.c |> %.d</code>

<td>SyntaxError

<tr><td colspan="2"><strong>Assignment</strong>

<tr>
<td><code>a = b |> %.c |> %.d</code>

<td><code>a = (b |> %.c |> %.d)</code>

<tr>
<td><code>a |> (b = %.c) |> %.d</code>

<td><code>a |> (b = %.c) |> %.d</code>

<tr>
<td><code>a |> (b = %.c |> %.d)</code>

<td><code>a |> (b = (%.c |> %.d))</code>

<tr>
<td><code>a |> b = %.c |> %.d</code>

<td>SyntaxError

<tr><td colspan="2"><strong>Conditionals</strong>

<tr>
<td><code>a ? b : c |> %.d |> %.e</code>

<td><code>a ? b : (c |> %.d |> %.e)</code>

<tr>
<td><code>a ? b |> %.c : d |> %.e</code>

<td><code>a ? (b |> %.c) : (d |> %.e)</code>

<tr>
<td><code>(a |> %.b) ? c : d |> %.e</code>

<td><code>(a |> %.b) ? c : (d |> %.e)</code>

<tr>
<td><code>a |> (%.b ? c : d) |> %.e</code>

<td><code>a |> (%.b ? c : d) |> %.e</code>

<tr>
<td><code>a |> %.b ? c : d |> %.e</code>

<td>SyntaxError

<tr><td colspan="2"><strong>Yields</strong>

<tr>
<td><code>yield a |> %.b</code>

<td><code>yield (a |> %.b)</code>

<tr>
<td><code>a |> (yield %) |> %.b</code>

<td><code>a |> (yield %) |> %.b</code>

<tr>
<td><code>a |> (yield % |> %.b)</code>

<td><code>a |> (yield (% |> %.b))</code>

<tr>
<td><code>a |> yield % |> %.b</code>

<td>SyntaxError

</table>

**Comment by js-choi** [comment]
This was addressed in 6c458b5. A `|>` PipeExpression now has ShortCircuitExpression on its left and AssignmentExpression on its right, and the right-hand side must not cover an unparenthesized operation with the same precedence (ArrowFunction, AsyncArrowFunction, YieldExpression, ConditionalExpression, or an assignment operation).

The [specification](https://jschoi.org/21/es-hack-pipes/) has been updated.
